minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/util/CommandExecutor.java

```diff
@@ -0,0 +1,163 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.minikube.util;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.gradle.api.GradleException;
+import org.gradle.api.logging.Logger;
+import org.gradle.internal.impldep.com.google.common.annotations.VisibleForTesting;
+
+/** Executes a shell command. */
+public class CommandExecutor {
+  public CommandExecutor() {}
+
+  public CommandExecutor setLogger(Logger logger) {
+    this.logger = logger;
+    return this;
+  }
+
+  @VisibleForTesting static final int TIMEOUT_SECONDS = 5;
+
+  @VisibleForTesting
+  static class ProcessBuilderFactory {
+    ProcessBuilder createProcessBuilder() {
+      return new ProcessBuilder();
+    }
+  }
+
+  @VisibleForTesting
+  CommandExecutor setProcessBuilderFactory(ProcessBuilderFactory processBuilderFactory) {
+    this.processBuilderFactory = processBuilderFactory;
+    return this;
+  }
+
+  @VisibleForTesting
+  static class ExecutorServiceFactory {
+    ExecutorService createExecutorService() {
+      return Executors.newSingleThreadExecutor();
+    }
+  }
+
+  @VisibleForTesting
+  CommandExecutor setExecutorServiceFactory(ExecutorServiceFactory executorServiceFactory) {
+    this.executorServiceFactory = executorServiceFactory;
+    return this;
+  }
+
+  private ProcessBuilderFactory processBuilderFactory = new ProcessBuilderFactory();
+  private ExecutorServiceFactory executorServiceFactory = new ExecutorServiceFactory();
+  private Logger logger;
+
+  /**
+   * Runs the command.
+   *
+   * @param command the list of command line tokens
+   * @return the output of the command as a list of lines
+   * @throws GradleException if the command exited with non-zero exit code
+   */
+  public List<String> run(List<String> command) throws IOException, InterruptedException {
+    if (logger != null) {
+      logger.debug("Running command : " + String.join(" ", command));
+    }
+
+    ExecutorService executor = executorServiceFactory.createExecutorService();
+
+    // Builds the command to execute.
+    final Process process = buildProcess(command);
+
+    // Runs the command and streams the output.
+    List<String> output = new ArrayList<>();
+    int exitCode = runProcessWithExecutor(process, executor, output);
+    shutdownExecutor(executor);
+
+    // Stops the build if the command fails to do something, we may want to make this configurable.
+    if (exitCode != 0) {
+      throw new GradleException("command exited with non-zero exit code : " + exitCode);
+    }
+
+    return output;
+  }
+
+  private Process buildProcess(List<String> command) throws IOException {
```

Comment:
Can probably just leave in `run`. We're not testing this anyway, it feels slightly harder to follow.